cpp/src/util/file_io.hpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include <raft/core/error.hpp>
+
+#include <algorithm>
+#include <cstring>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+namespace cuvs::util {
+
+/**
+ * @brief RAII wrapper for POSIX file descriptors
+ *
+ * Manages file descriptor lifecycle with automatic cleanup.
+ * Non-copyable, move-only.
+ */
+class file_descriptor {
+ public:
+  explicit file_descriptor(int fd = -1) : fd_(fd) {}
+
+  file_descriptor(const std::string& path, int flags, mode_t mode = 0644)
+    : fd_(open(path.c_str(), flags, mode))
+  {
+    if (fd_ == -1) {
+      RAFT_FAIL("Failed to open file: %s (errno: %d, %s)", path.c_str(), errno, strerror(errno));
+    }
+  }
+
+  file_descriptor(const file_descriptor&)            = delete;
+  file_descriptor& operator=(const file_descriptor&) = delete;
+
+  file_descriptor(file_descriptor&& other) noexcept : fd_(other.fd_) { other.fd_ = -1; }
+
+  file_descriptor& operator=(file_descriptor&& other) noexcept
+  {
+    if (this != &other) {
+      close();
+      fd_       = other.fd_;
+      other.fd_ = -1;
+    }
+    return *this;
+  }
```

Comment:
A nitpick:
```suggestion
  {
    std::swap(this->fd_, other.fd_);
    return *this;
  }
```
Note you don't need to manually close the handle here; the destructor will be called on the moved-from object anyway.

Reply:
Applied.